Magento/Sniffs/Classes/ConstructorProxyInterceptorSniff.php

```diff
@@ -0,0 +1,267 @@
+<?php
+/**
+ * Copyright © Magento. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Sniffs\Classes;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Detects explicit request of proxies and interceptors in constructors
+ *
+ * Search is in variable names and namespaces, including indirect namespaces from use statements
+ */
+class ConstructorProxyInterceptorSniff implements Sniff
+{
+    const CONSTRUCT_METHOD_NAME = '__construct';
+
+    /**
+     * String representation of warning.
+     *
+     * @var string
+     */
+    protected $warningMessage = 'Proxies and interceptors MUST never be explicitly requested in constructors.';
+
+    /**
+     * Warning violation code.
+     *
+     * @var string
+     */
+    protected $warningCode = 'ConstructorProxyInterceptor';
+
+    /**
+     * @var array Aliases of proxies or plugins from use statements
+     */
+    private $aliases = [];
+
+    /**
+     * @var array Terms to search for in variables and namespaces
+     */
+    private $warnNames = [
+        'proxy',
+        'plugin'
```

Comment:
Probably the task was not clear enough, the expectation was to prevent `Interceptor` classes being referenced directly. Interceptors are auto-generated classes. 
It may make sense to run these tests against Magento codebase and see if Plugin reference detection also makes sense.

Reply:
Makes sense. I'll add interceptors first and then look at plugins as you suggest.